pkg/security/spiffe/spiffe.go

```diff
@@ -24,42 +26,77 @@ import (
 
 // Parsed is a parsed SPIFFE ID according to the Dapr SPIFFE ID path format.
 type Parsed struct {
-	TrustDomain string
-	Namespace   string
-	AppID       string
+	id spiffeid.ID
+
+	namespace string
+	appID     string
 }
 
-// FromContext parses a SPIFFE ID from a gRPC context.
-func FromContext(ctx context.Context) (Parsed, bool, error) {
+// FromGRPCContext parses a SPIFFE ID from a gRPC context.
+func FromGRPCContext(ctx context.Context) (*Parsed, bool, error) {
 	// Apply access control list filter
 	id, ok := grpccredentials.PeerIDFromContext(ctx)
 	if !ok {
-		return Parsed{}, false, nil
+		return nil, false, nil
 	}
 
-	return fromID(id)
-}
-
-func fromID(id spiffeid.ID) (Parsed, bool, error) {
 	split := strings.Split(id.Path(), "/")
 	// Don't force match of 4 segments, since we may want to add more identifiers
 	// to the path in future which would otherwise break backwards compat.
 	if len(split) < 4 || split[0] != "" || split[1] != "ns" {
-		return Parsed{}, false, fmt.Errorf("malformed SPIFFE ID: %s", id.String())
+		return nil, false, fmt.Errorf("malformed SPIFFE ID: %s", id.String())
 	}
 
-	return Parsed{
-		TrustDomain: id.TrustDomain().String(),
-		Namespace:   split[2],
-		AppID:       split[3],
+	return &Parsed{
+		id:        id,
+		namespace: split[2],
+		appID:     split[3],
 	}, true, nil
 }
 
-func (p Parsed) ToID() (spiffeid.ID, error) {
-	td, err := spiffeid.TrustDomainFromString(p.TrustDomain)
+// FromStrings builds a Dapr SPIFFE ID with the given namespace and app ID in
+// the given Trust Domain.
+func FromStrings(td spiffeid.TrustDomain, namespace, appID string) (*Parsed, error) {
+	if len(td.String()) == 0 || len(namespace) == 0 || len(appID) == 0 {
+		return nil, errors.New("malformed SPIFFE ID")
+	}
+
+	id, err := spiffeid.FromSegments(td, "ns", namespace, appID)
 	if err != nil {
-		return spiffeid.ID{}, err
+		return nil, err
 	}
 
-	return spiffeid.FromSegments(td, "ns", p.Namespace, p.AppID)
+	return &Parsed{
+		id:        id,
+		namespace: namespace,
+		appID:     appID,
+	}, nil
+}
+
+func (p *Parsed) TrustDomain() spiffeid.TrustDomain {
```

Comment:
All of these methods would probably be not necessary if we did not change From* functions to return a pointer. Which would also cause less allocations on the heap.

Reply:
We don't want to make `Parsed` mutable once it has been created is the reason for creating these functions.

Reply:
Ok